### PR TITLE
Initial GraphPresenter prototype for #17

### LINKED ISSUE #17: [Feature] Add GraphPresenter to help with general calls to the Graph
## Describe the problem this feature would solve
Analogous to [MGT Issue #160 here](https://github.com/microsoftgraph/microsoft-graph-toolkit/issues/160)

Have a generalized component to aid in making graph calls within XAML for controls that we don't have yet.

## Describe the solution
Create a `GraphPresenter` control which inherits from `ContentPresenter`. By providing it a call to the graph, it will go fetch data from the graph and set its contents to the result.

This will allow the developer to setup a `DataTemplate` in order to display the proper data.

## Additional context & Screenshots

Screenshot of initial prototype:
image

Advantage of this proposed approach is that everything is still strongly typed to the Graph .NET SDK. 🦙

### LINKED ISSUE #27: MockProvider can't retrieve CalendarView events
## Describe the bug
Calling this fails with MockProvider, works fine with InteractiveProvider:
```
                // Calendar Data, Today and Tomorrow (As Local)
                var today = DateTimeOffset.Now.Date.ToUniversalTime();
                var events = await provider.Graph.Me.CalendarView.Request(new[]
                {
                    new QueryOption("startDateTime", today.ToString("o", CultureInfo.InvariantCulture)),
                    new QueryOption("endDateTime", today.AddDays(2).ToString("o", CultureInfo.InvariantCulture)),
                }).OrderBy("start/dateTime").GetAsync();
```

## Expected behavior
Should retrieve events, see #24, seems like there's an issue in our workaround from #25 in order to get around the Graph SDK issue: https://github.com/microsoftgraph/msgraph-sdk-dotnet-core/issues/59

### PR DESCRIPTION
**Pull request by michael-hawker**
Fixes #17 
Fixes #27 

Pushed some things needed here to the Toolkit in [https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3033](https://github.com/windows-toolkit/WindowsCommunityToolkit/pull/3033), will update this once those get merged in.

'Done':
- [ ] ~~How to deal with paging? Use IncrementalLoadingCollection?~~ Opened Issue #54 to track for future
- [x] How to deal with QueryOptions? Can we take in XAML???
- [ ] ~~Sample: What to do with empty thumbnails, should the converter just be a 'sample' one vs. part of the lib?~~ Needs https://github.com/microsoftgraph/MSGraph-SDK-Code-Generator/issues/266 to be fixed to continue OneDrive sample
- [ ] ~~XAML Islands: Test in XAML Islands, maybe make it so XAML Islands UWP just uses MainPage from UWP sample???~~ Doesn't need to be part of this work
- [x] How do we order in reverse in Graph API??? (add 'desc' with a space before to the OrderBy field) [See here](https://docs.microsoft.com/en-us/graph/query-parameters#orderby-parameter)

TODO/Questions:
- [ ] Test: Make sure getting the right number of days in CalendarView sample
- [x] Clean-up HTML output in Teams message sample
- [x] Move FileSize Converter to Toolkit? (not worry about for now, can do later)
- [x] Remove use of NotifyTaskCompletion (is that sample only use?) Done and added work-around to #55

### CODE REVIEW

**Comment by michael-hawker** [comment]
Need to fix StyleCop issues:
```
         Extensions\GraphExtensions.cs(183,1): error SA1137: Elements should have the same indentation [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(181,47): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(182,38): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(154,32): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(158,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(162,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(166,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(170,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]
         Extensions\GraphExtensions.cs(174,40): error SA1139: Use literal suffix notation instead of casting [D:\a\1\s\Microsoft.Toolkit.Graph\Microsoft.Toolkit.Graph.csproj]

```

**Comment by michael-hawker** [comment]
```
         Converters\FileSizeConverter.cs(11,18): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Converters\FileSizeConverter.cs(13,23): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter.Convert(object, Type, object, string)' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Converters\FileSizeConverter.cs(23,23): error CS1591: Missing XML comment for publicly visible type or member 'FileSizeConverter.ConvertBack(object, Type, object, string)' [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Providers\NotifyTaskCompletion.cs(141,45): error SA1502: Element should not be on a single line [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
         Providers\NotifyTaskCompletion.cs(109,36): error SA1502: Element should not be on a single line [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj
```

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+                // Handle Special QueryOptions
+                if (!string.IsNullOrWhiteSpace(OrderBy))
+                {
+                    request.QueryOptions.Add(new Microsoft.Graph.QueryOption("$orderby", OrderBy));
+                }
+
+                // TODO: Add Exception Handling
+                // Note: CalendarView not supported https://github.com/microsoftgraph/msgraph-sdk-dotnet/issues/740
+                var response = await request.SendAsync<object>(null, CancellationToken.None).ConfigureAwait(false) as JObject;
```
[the remark this answers is not in the document]
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:93]: Thanks Sergio, I did mean to do something here. We should probably have some thoughts in general about loading/error states for the Graph controls in general. I should at least wrap and ignore this for now though. :)

File: Microsoft.Toolkit.Graph/Extensions/GraphExtensions.cs [hunk 2]
```diff
@@ -136,5 +137,42 @@ public static T Search<T>(this T request, string query)
 
             return request;
         }
+
+        /// <summary>
+        /// Translate numeric file size to string format.
+        /// </summary>
+        /// <param name="size">file size in bytes.</param>
+        /// <returns>Returns file size string.</returns>
+        public static string ToFileSizeString(this long size) // TODO: Move this to Microsoft.Toolkit
```
[the remark this answers is not in the document]
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph/Extensions/GraphExtensions.cs:146]: I completely forgot about it, I should have a reference too, I'll remove this from here. Thanks!

File: SampleTest/MainPage.xaml [hunk 3]
```diff
@@ -25,19 +28,236 @@
...
+                                    IsCollection="True">
+                                    <wgt:GraphPresenter.ContentTemplate>
+                                        <DataTemplate>
+                                            <ScrollViewer HorizontalScrollMode="Disabled" VerticalScrollBarVisibility="Auto">
+                                                <ItemsControl ItemsSource="{Binding}">
+                                                    <ItemsControl.ItemTemplate>
+                                                        <DataTemplate x:DataType="graph:Message">
+                                                            <StackPanel>
+                                                                <wgt:PersonView PersonQuery="{x:Bind Sender.EmailAddress.Address}" ShowName="True" Margin="-4"/>
+                                                                <TextBlock Text="{Binding Subject}" Style="{StaticResource BaseTextBlockStyle}" Padding="0"/>
```
[the remark this answers is not in the document]
Comment (michael-hawker) [reply SampleTest/MainPage.xaml:142]: Just forgot to update it from when I was playing around the first time :)

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+                }
+                else
+                {
+                    data = values.ToObject(ResponseType);
+                }
+
+                await DispatcherHelper.ExecuteOnUIThreadAsync(() =>
+                {
+                    Content = data;
+                });
```
[the remark this answers is not in the document]
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs]: Ordering was odd, so couldn't apply this directly, but resolved now.

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+        }
+
+        /// <summary>
+        /// Identifies the <see cref="RequestBuilder"/> dependency property.
+        /// </summary>
+        /// <returns>
+        /// The identifier for the <see cref="RequestBuilder"/> dependency property.
+        /// </returns>
+        public static readonly DependencyProperty RequestBuilderProperty =
+            DependencyProperty.Register(nameof(RequestBuilder), typeof(IBaseRequestBuilder), typeof(GraphPresenter), new PropertyMetadata(null));
```
Comment (azchohfi) [inline Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:38]: What if the RequestBuilder changes after the GraphPresenter is loaded? Shouldn't this be handled on the DP's changed event?

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+
+        /// <summary>
+        /// Gets or sets list of <see cref="QueryOption"/> representing <see cref="Microsoft.Graph.QueryOption"/> values to pass into the request built by <see cref="RequestBuilder"/>.
+        /// </summary>
+        public List<QueryOption> QueryOptions { get; set; } = new List<QueryOption>();
+
+        /// <summary>
+        /// Gets or sets a string to indicate a sorting order for the <see cref="RequestBuilder"/>. This is a helper to add this specific request option to the <see cref="QueryOptions"/>.
+        /// </summary>
+        public string OrderBy { get; set; }
```
Comment (azchohfi) [inline Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:59]: Wouldn't it make sense to make all these properties as DPs, and react (do the query again with the updated parameters) when they change?

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+                    if (IsCollection)
+                    {
+                        data = values.ToObject(Array.CreateInstance(ResponseType, 0).GetType());
+                    }
+                    else
+                    {
+                        data = values.ToObject(ResponseType);
+                    }
+
+                    _ = DispatcherHelper.ExecuteOnUIThreadAsync(() => Content = data);
```
Comment (azchohfi) [inline Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:104]: Use the DispatcherQueue, not the old CoreDispatcher.

File: SampleTest/MainPage.xaml [hunk 8]
```diff
@@ -2,16 +2,19 @@
...
     xmlns:graph="using:Microsoft.Graph"
+    xmlns:global="using:System.Globalization"
     xmlns:Interactivity="using:Microsoft.Xaml.Interactivity"
     xmlns:providers="using:Microsoft.Toolkit.Graph.Providers"
+    xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"
+    xmlns:ex="using:Microsoft.Toolkit.Extensions"
     mc:Ignorable="d"
     Background="{ThemeResource ApplicationPageBackgroundThemeBrush}"
-    xmlns:controls="using:Microsoft.Toolkit.Uwp.UI.Controls"><!-- Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/1286 -->
+    xmlns:controlsKeepThis="using:Microsoft.Toolkit.Uwp.UI.Controls"><!-- Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/1286 -->
```
Comment (azchohfi) [inline SampleTest/MainPage.xaml:17]: If we bump the target SDK, should this issue go away? If so, we should do this.

Comment (azchohfi) [reply SampleTest/MainPage.xaml:17]: Also, this is being added right here on the top(line 13), so I don't think you need it anymore.

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+using System.Text;
+using System.Threading.Tasks;
+using Windows.UI.Xaml;
+
+namespace Microsoft.Toolkit.Graph.Controls
+{
+    /// <summary>
+    /// XAML Proxy for <see cref="Microsoft.Graph.QueryOption"/>.
+    /// </summary>
+    public sealed class QueryOption
```
Comment (azchohfi) [inline Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:17]: Shouldn't we reuse the existing ListDictionary and keep from adding another Key<->Value mapping class?
Something like this:
```xaml
...
xmlns:specialized="using:System.Collections.Specialized"
...
<specialized:ListDictionary>
    <x:String x:Key="key1">Value1</x:String>
    <x:String x:Key="key2">Value2</x:String>
</specialized:ListDictionary>
```
One less class to maintain.

File: Microsoft.Toolkit.Graph/Providers/MockProvider.cs [hunk 10]
```diff
@@ -42,22 +44,15 @@ private set
...
-                //// End Workaround
-
-                return this.AuthenticateRequestAsync(requestMessage);
-            }));
+                        new DelegateAuthenticationProvider((requestMessage) =>
+                    {
+                        var requestUri = requestMessage.RequestUri.ToString();
+
+                        // Prepend Proxy Service URI to our request
+                        requestMessage.RequestUri = new Uri(GRAPH_PROXY_URL + HttpUtility.UrlEncode(requestUri));
```
Comment (michael-hawker) [inline Microsoft.Toolkit.Graph/Providers/MockProvider.cs:52]:
```suggestion
                        requestMessage.RequestUri = new Uri(GRAPH_PROXY_URL + Uri.EscapeDataString(requestUri));
```

Looks like we can use `Uri.EscapeDataString` here instead which reduces one more reference (System.Web) and is more compatible with the Uno build.

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs (hunk 7 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:104]: We're not on the 7.0.0 preview package yet. I need to update later.

File: SampleTest/MainPage.xaml.cs [hunk 11]
```diff
@@ -11,9 +16,44 @@ namespace SampleTest
...
+        public DateTime ThreeDaysFromNow => Today.AddDays(3);
+
         public MainPage()
         {
             this.InitializeComponent();
         }
+
+        public static string ToLocalTime(DateTimeTimeZone value)
+        {
+            //// Workaround for https://github.com/microsoft/microsoft-ui-xaml/issues/2407
```
[the remark this answers is not in the document]
Comment (michael-hawker) [reply SampleTest/MainPage.xaml.cs:32]: Made `//` instead

File: SampleTest/MainPage.xaml (hunk 8 above)
Comment (michael-hawker) [reply SampleTest/MainPage.xaml:17]: I might bump/test that as a separate PR?

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs (hunk 9 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:17]: I didn't know this existed 🙂. I do like having close mapping to the intent though for the Graph. And we'd lose the implicit operator converter, eh? Let's chat quick in the morning.

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs (hunk 6 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:59]: For this and the RequestBuilder, I didn't expect them to change much yet. I think that's a separate PR to add that functionality in?

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs (hunk 9 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:17]: I don't know how we bind to Value in these cases, so let's leave this for now and explore more later. I think there does need to be some refactoring to ensure timings and/or dynamic updates to query options...

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs (hunk 6 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:59]: Opened issue https://github.com/windows-toolkit/Graph-Controls/issues/56 to track this for later.

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs (hunk 5 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:38]: https://github.com/windows-toolkit/Graph-Controls/issues/56

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs (hunk 7 above)
Comment (michael-hawker) [reply Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/GraphPresenter.cs:104]: #57

File: SampleTest/MainPage.xaml (hunk 8 above)
Comment (michael-hawker) [reply SampleTest/MainPage.xaml:17]: Added to #57

**Comment by michael-hawker** [comment]
@azchohfi thanks for the review, addressed your comments as well and opened #56 for tracking. If it looks good, please merge this.

Also added #57 for updating to our new 7.0.0-preview2 package and the Target SDK bump. If you have a chance to do that update today, that'd be big help for UnoConf. Then we could update our preview package for this as well to align with `7.0.0-preview2`.

Thanks! 🦙❤

**Comment by azchohfi** [comment]
Working on it!
Small issue!
```
         Controls\GraphPresenter\QueryOption.cs(28,57): error SA1629: Documentation text should end with a period [D:\a\1\s\Microsoft.Toolkit.Graph.Controls\Microsoft.Toolkit.Graph.Controls.csproj]
```

File: Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+        /// <inheritdoc cref="Microsoft.Graph.Option.Name"/>
+        public string Name { get; set; }
+
+        /// <inheritdoc cref="Microsoft.Graph.Option.Value"/>
+        public string Value { get; set; }
+
+        /// <summary>
+        /// Implicit conversion for <see cref="QueryOption"/> to <see cref="Microsoft.Graph.QueryOption"/>.
+        /// </summary>
+        /// <param name="option">query option to convert</param>
```
Comment (azchohfi) [inline Microsoft.Toolkit.Graph.Controls/Controls/GraphPresenter/QueryOption.cs:28]:
```suggestion
        /// <param name="option">query option to convert.</param>
```

**Pull request merged**